src/wallet/sqlite.h

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_WALLET_SQLITE_H
+#define BITCOIN_WALLET_SQLITE_H
+
+#include <wallet/db.h>
+
+struct bilingual_str;
+class SQLiteDatabase;
+
+/** RAII class that provides access to a WalletDatabase */
+class SQLiteBatch : public DatabaseBatch
+{
+private:
+    SQLiteDatabase& m_database;
+
+    bool m_read_only = false;
+
+    bool ReadKey(CDataStream&& key, CDataStream& value) override;
+    bool WriteKey(CDataStream&& key, CDataStream&& value, bool overwrite=true) override;
+    bool EraseKey(CDataStream&& key) override;
+    bool HasKey(CDataStream&& key) override;
+
+public:
+    explicit SQLiteBatch(SQLiteDatabase& database, const char* mode);
+
+    /* No-op. See commeng on SQLiteDatabase::Flush */
+    void Flush() override {}
+
+    void Close() override;
+
+    bool StartCursor() override;
+    bool ReadAtCursor(CDataStream& ssKey, CDataStream& ssValue, bool& complete) override;
+    void CloseCursor() override;
+    bool TxnBegin() override;
+    bool TxnCommit() override;
+    bool TxnAbort() override;
+};
+
+/** An instance of this class represents one SQLite3 database.
+ **/
+class SQLiteDatabase : public WalletDatabase
+{
+private:
+    bool m_mock = false;
+
+    const std::string m_dir_path;
+
+    const std::string m_file_path;
+
+public:
+    SQLiteDatabase() = delete;
+
+    /** Create DB handle to real database */
+    SQLiteDatabase(const fs::path& dir_path, const fs::path& file_path, bool mock=false);
+
+    ~SQLiteDatabase();
+
+    /** Open the database if it is not already opened */
+    void Open(const char* mode) override;
+
+    /** Close the database */
+    void Close() override;
+
+    /* These functions are unused */
+    void AddRef() override { assert(false); }
+    void RemoveRef() override { assert(false); }
+
+    /** Rewrite the entire database on disk */
+    bool Rewrite(const char* skip=nullptr) override;
```

Comment:
1a2eec03f0dc7731d5c199a97db148534ce174a2, nit -- style:
```suggestion
    bool Rewrite(const char* skip = nullptr) override;
```

Reply:
Done